test/rbacHierarchicalDomain.test.ts

```diff
@@ -0,0 +1,98 @@
+    await e.addNamedGroupingPolicy('g2', 'domain2', 'domain3');
+    (e.getRoleManager() as DefaultRoleManager).addDomainHierarchy(e.getNamedRoleManager('g2') as RoleManager);
+  });
+
+  test('Authorization to lower domain should pass', async () => {
+    expect(e.enforceSync('alice', 'domain1', 'data1', 'read')).toBe(true);
+    expect(e.enforceSync('alice', 'domain1', 'data2', 'read')).toBe(true);
+    expect(e.enforceSync('alice', 'domain2', 'data2', 'read')).toBe(true);
+    expect(e.enforceSync('alice', 'domain1', 'sdata2', 'read')).toBe(true);
+    expect(e.enforceSync('alice', 'sibling2', 'sdata2', 'read')).toBe(true);
+    expect(e.enforceSync('alice', 'domain3', 'data3', 'write')).toBe(true);
+    expect(e.enforceSync('alice', 'domain1', 'data3', 'write')).toBe(true);
+    expect(e.enforceSync('alice', 'domain2', 'data3', 'write')).toBe(true);
+    expect(e.enforceSync('bob', 'domain2', 'data2', 'read')).toBe(true);
+    expect(e.enforceSync('bob', 'domain2', 'data3', 'write')).toBe(true);
+    expect(e.enforceSync('bob', 'domain3', 'data3', 'write')).toBe(true);
+  });
+
+  test('Authorization to lower domain should faill if no role', async () => {
+    expect(e.enforceSync('bob', 'domain1', 'data3', 'write')).toBe(false);
+    expect(e.enforceSync('bob', 'domain1', 'data2', 'read')).toBe(false);
+  });
+
+  test('Authorization to higher domain without permissions should faill', async () => {
+    expect(e.enforceSync('bob', 'domain1', 'data1', 'read')).toBe(false);
+    expect(e.enforceSync('bob', 'domain2', 'data1', 'read')).toBe(false);
+    expect(e.enforceSync('bob', 'domain3', 'data1', 'write')).toBe(false);
+  });
+
+  test('Authorization to higher domain with permissions should faill', async () => {
+    // alice data1
+    expect(e.enforceSync('alice', 'domain2', 'data1', 'read')).toBe(false);
+    expect(e.enforceSync('alice', 'domain3', 'data1', 'read')).toBe(false);
+    expect(e.enforceSync('alice', 'sibling2', 'data1', 'read')).toBe(false);
+    // alice data2
+    expect(e.enforceSync('alice', 'domain3', 'data2', 'read')).toBe(false);
+    expect(e.enforceSync('alice', 'sibling2', 'data2', 'read')).toBe(false);
+    // alice data3
+    expect(e.enforceSync('alice', 'sibling2', 'data3', 'write')).toBe(false);
+    // alice sdata2
+    expect(e.enforceSync('alice', 'domain2', 'sdata2', 'read')).toBe(false);
+    // bob data2
+    expect(e.enforceSync('bob', 'domain3', 'data2', 'read')).toBe(false);
+  });
+
+  test('Authorization to sibling without permissions should faill', async () => {
+    expect(e.enforceSync('bob', 'domain1', 'sdata2', 'read')).toBe(false);
+    expect(e.enforceSync('bob', 'domain2', 'sdata2', 'read')).toBe(false);
+    expect(e.enforceSync('bob', 'sibling2', 'sdata2', 'read')).toBe(false);
+    expect(e.enforceSync('bob', 'domain3', 'sdata2', 'read')).toBe(false);
+  });
+
+  test('User should not have a permission if there is none in the lower domain', async () => {
```

Comment:
`User should not have a permission if there is none in the lower domain`
I have a question,
what if we want to add like a super-admin type of role, and we want it to have permission in lower domain also
like if `alice` is `admin` in the highest domain then `alice` should be `admin` in the lower domain also.
for this type of hierarchy, writing grouping policy for each domain is the only option?

Reply:
In this case you would need to create a role in every domain yeah. Or just inherit roles in upper domain to another role. Like alice is technically a role and an user.

There are other options as well. You could modify the model a little bit by adding regex to the object.

Then the role could be just `superadmin, (.*), (.*)` which has access to everything, but at that point the rights come from regex and not hierarchical domain :)

Reply:
ok :+1: